e2e: auto-detect running on OpenShift

nixpanic opened the pull request:
The e2e suite can run on OpenShift, but there are a few checks that
handle certain things with a small change. OpenShift has a more struct
security policy than standard Kubernetes/Minikube, these need to be
accounted for.

There used to be a -is-openshift commandline flag. This flag is now
removed and OpenShift is auto-detected.

This PR also contains fixes for running on Kubernetes 1.34.

nixpanic committed:
e2e: auto-detect running on OpenShift

The e2e suite can run on OpenShift, but there are a few checks that
handle certain things with a small change. OpenShift has a more struct
security policy than standard Kubernetes/Minikube, these need to be
accounted for.

There used to be a -is-openshift commandline flag. This flag is now
removed and OpenShift is auto-detected.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

Madhu-1 commented on e2e/openshift.go:
```diff
@@ -0,0 +1,45 @@
+// If the clusterversion is not found, we're not running on an OpenShift
+// cluster.
+func detectOpenShift() (bool, error) {
+	version, err := e2ekubectl.RunKubectl("openshift",
```
does this work.  i thought it always run kubectl and no other binaries? do we have sample output of this one?

nixpanic replied:
Yes, like this:
```
$ ../e2e.test -test-cephfs=false -test-nfs=false -test-rbd=false -test-nbd=false -ginkgo.v=true
  I1002 10:37:51.596865 241957 test_context.go:564] The --provider flag is not set. Continuing as if --provider=skeleton had been used.
timeout for deploytimeout  10
  I1002 10:37:51.597042 241957 builder.go:121] Running '/home/ndevos/bin/kubectl --kubeconfig=/tmp/kubeconfig --namespace=openshift get clusterversion.config.openshift.io/version -ojsonpath='{.status.desired.version}''
  I1002 10:37:52.884262 241957 builder.go:146] stderr: ""
  I1002 10:37:52.884351 241957 builder.go:147] stdout: "'4.19.14'"
  I1002 10:37:52.884385 241957 openshift.go:42] running on OpenShift version '4.19.14'
Running Suite: E2e Suite - /home/ndevos/go/src/github.com/ceph/ceph-csi/e2e
===========================================================================
Random Seed: 1759394271

Will run 0 of 0 specs

Ran 0 of 0 Specs in 0.000 seconds
SUCCESS! -- 0 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
```

If I intentionally break it with an invalid kind (like on standard Kubernetes), it looks like this:
```
$ ../e2e.test -test-cephfs=false -test-nfs=false -test-rbd=false -test-nbd=false -ginkgo.v=true
  I1002 10:36:09.224606 241653 test_context.go:564] The --provider flag is not set. Continuing as if --provider=skeleton had been used.
timeout for deploytimeout  10
  I1002 10:36:09.224770 241653 builder.go:121] Running '/home/ndevos/bin/kubectl --kubeconfig=/tmp/kubeconfig --namespace=openshift get clusterversion.config.openshift.XXio/version -ojsonpath='{.status.desired.version}''
  I1002 10:36:10.203555 241653 builder.go:135] rc: 1
Running Suite: E2e Suite - /home/ndevos/go/src/github.com/ceph/ceph-csi/e2e
===========================================================================
Random Seed: 1759394169

Will run 0 of 0 specs

Ran 0 of 0 Specs in 0.000 seconds
SUCCESS! -- 0 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
```

Madhu-1 replied:
ha okay, i thought its a binary name , forgot its a namespace name 🤦🏻

nixpanic replied:
Right. The 1st parameters is the Namespace. The clusterversion is actually a cluster-scoped CRD. The namespace argument does not seem to change the behavior.

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.34

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.33

nixpanic commented:
> /test ci/centos/mini-e2e/k8s-1.34

This one failed without clear cause to me. [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/job/mini-e2e_k8s-1.34/8/display/redirect)

nixpanic committed:
e2e: use test framework logging, not fmt.Println

By using the `framework.Logf()` function, the format of the logs are
more structured. All messages get the same format and time prefixed.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
Added a minor logging change that annoyed me while running tests.

nixpanic commented:
/test ci/centos/mini-e2e/k8s-1.33

Rakshith-R commented:
@mergifyio rebase

Rakshith-R commented:
@mergifyio queue

nixpanic commented:
@mergifyio rebase

nixpanic commented:
@mergifyio queue

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.34

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.32

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.34

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.33

nixpanic commented:
Retriggered CI jobs because they failed to checkout the GitHub repository.

nixpanic commented:
/retest ci/centos/upgrade-tests-rbd

nixpanic committed:
e2e: detect updated error message for RWOP test

On Kubernetes 1.34 there is a new error returned during the RWOP test:

> unavailable due to PersistentVolumeClaim with ReadWriteOncePod access
> mode already in-use by another pod

Because this error differs from the expected RWOP failure in previous
Kubernetes versions, the test fails.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.34

nixpanic committed:
e2e: skip detecting metrics on Kubernetes 1.34

kubelet_volume_stats_* are not available on Kubernetes 1.34. CI testing
fails because `validateNormalUserPVCAccessFunc()` check them.

See-also: kubernetes/kubernetes#133847
Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.34

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.34

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.34

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.34

nixpanic committed:
e2e: improve timeout and podname in logging message

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.34

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.34

Madhu-1 commented on e2e/utils.go:
```diff
@@ -1688,6 +1685,25 @@ func k8sVersionGreaterEquals(c kubernetes.Interface, major, minor int) bool {
+	if brokenMetrics == nil {
+		broken := true
+		functional := true
+		if k8sVersionGreaterEquals(c, 1, 34) && !k8sVersionGreaterEquals(c, 1, 35) {
```
lets open a tracker to revert it back even if its fixed in some patch 1.34 release

nixpanic replied:
Done with #5635.

nixpanic commented:
@mergifyio queue
